### PR TITLE
CPM-560: Add group associations user intents

### PR DESCRIPTION
**Pull request by cemorin**
**Description (for Contributor and Core Developer)**
Add following user intents:
- AssociateGroups
- DissociateGroups
- ReplaceGroups

**Definition Of Done (for Core Developer only)**

- [x] Tests
- [ ] Migration & Installer
- [ ] PM Validation (Story)
- [ ] Changelog (maintenance bug fixes)
- [ ] Tech Doc

### CODE REVIEW

File: src/Akeneo/Pim/Enrichment/Product/back/Application/Applier/AssociationUserIntentCollectionApplier.php [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+                        \array_merge($formerAssociations, $associationUserIntent->groupCodes())
+                    )
+                );
+            } elseif ($associationUserIntent instanceof DissociateGroups) {
+                $newAssociations = \array_diff($formerAssociations, $associationUserIntent->groupCodes());
+                if (\count($newAssociations) === \count($formerAssociations)) {
+                    continue;
+                }
+
+                $normalizedAssociations[$associationUserIntent->associationType()]['groups'] = \array_values(
```
Comment (cemorin) [inline src/Akeneo/Pim/Enrichment/Product/back/Application/Applier/AssociationUserIntentCollectionApplier.php:83]: regroup associate, dissociate and replace + function to compute key array (group, products or product_model)

File: src/Akeneo/Pim/Enrichment/Product/back/Application/Applier/AssociationUserIntentCollectionApplier.php [hunk 2]
```diff
@@ -39,72 +47,23 @@ public function apply(UserIntent $userIntent, ProductInterface $product, int $us
         $normalizedAssociations = [];
 
         foreach ($userIntent->associationUserIntents() as $associationUserIntent) {
-            $formerAssociations = $this->getFormerAssociations($associationUserIntent, $normalizedAssociations, $product);
+            $values = null;
```
Comment (nmarniesse) [inline src/Akeneo/Pim/Enrichment/Product/back/Application/Applier/AssociationUserIntentCollectionApplier.php:50]: to remove?

File: src/Akeneo/Pim/Enrichment/Product/back/Application/Applier/AssociationUserIntentCollectionApplier.php [hunk 3]
```diff
@@ -39,72 +47,23 @@ public function apply(UserIntent $userIntent, ProductInterface $product, int $us
...
+            $formerAssociations = $this->getFormerAssociations($associationUserIntent, $normalizedAssociations, $product, $entityType);
+            $entityAssociations = $this->userIntentEntityAssociations($associationUserIntent, $entityType);
+
+            $values = match($associationUserIntent::class) {
+                AssociateProducts::class, AssociateProductModels::class, AssociateGroups::class =>
+                    $this->associateEntities($formerAssociations, $entityAssociations),
+                DissociateProducts::class, DissociateProductModels::class, DissociateGroups::class =>
+                    $this->dissociateEntities($formerAssociations, $entityAssociations),
+                ReplaceAssociatedProducts::class, ReplaceAssociatedProductModels::class, ReplaceAssociatedGroups::class =>
+                    $this->replaceAssociatedEntities($formerAssociations, $entityAssociations, $entityType, $userId)
```
Comment (nmarniesse) [inline src/Akeneo/Pim/Enrichment/Product/back/Application/Applier/AssociationUserIntentCollectionApplier.php:61]:
```suggestion
                    $this->replaceAssociatedEntities($formerAssociations, $entityAssociations, $entityType, $userId),
                    default => throw new \InvalidArgumentException('...')
```

File: src/Akeneo/Pim/Enrichment/Product/back/Application/Applier/AssociationUserIntentCollectionApplier.php [hunk 4]
```diff
@@ -124,49 +83,110 @@ public function getSupportedUserIntents(): array
...
+     */
+    private function userIntentEntityAssociations(AssociationUserIntent $associationUserIntent, string $entityType): array
+    {
+        if ($entityType === self::PRODUCTS && \method_exists($associationUserIntent, 'productIdentifiers')) {
+            return $associationUserIntent->productIdentifiers();
+        } elseif ($entityType === self::PRODUCT_MODELS && \method_exists($associationUserIntent, 'productModelCodes')) {
+            return $associationUserIntent->productModelCodes();
+        } elseif ($entityType === self::GROUPS && \method_exists($associationUserIntent, 'groupCodes')) {
+            return $associationUserIntent->groupCodes();
         }
```
Comment (nmarniesse) [inline src/Akeneo/Pim/Enrichment/Product/back/Application/Applier/AssociationUserIntentCollectionApplier.php:129]:
```suggestion
        if ($entityType === self::PRODUCTS) {
            return $associationUserIntent->productIdentifiers();
        } elseif ($entityType === self::PRODUCT_MODELS) {
            return $associationUserIntent->productModelCodes();
        } elseif ($entityType === self::GROUPS) {
            return $associationUserIntent->groupCodes();
        }
```

You can remove the method_exist. The code will also fail if the method does not exist

**Review by nmarniesse — APPROVED** [review]
GTM after fix

File: src/Akeneo/Pim/Enrichment/Product/back/Application/Applier/AssociationUserIntentCollectionApplier.php (hunk 4 above)
Comment (cemorin) [reply src/Akeneo/Pim/Enrichment/Product/back/Application/Applier/AssociationUserIntentCollectionApplier.php:129]: Without them phpStan throws an error, should I add php_stan ignore or keep the methods?

Comment (nmarniesse) [reply src/Akeneo/Pim/Enrichment/Product/back/Application/Applier/AssociationUserIntentCollectionApplier.php]: Ok keep like that

File: src/Akeneo/Pim/Enrichment/Product/back/Test/Integration/Handler/UpsertProductAssociationsIntegration.php [hunk 5]
```diff
@@ -205,4 +208,101 @@ public function it_replaces_product_model_associations(): void
         $updatedProduct = $this->productRepository->findOneByIdentifier('identifier');
         Assert::assertSame(['other_product_model'], $this->getAssociatedProductModelIdentifiers($updatedProduct));
     }
+
+    /**
+     * @return array<string>
+     */
+    public function getAssociatedProductModelIdentifiers(ProductInterface $product): array
```
Comment (mmetayer) [inline src/Akeneo/Pim/Enrichment/Product/back/Test/Integration/Handler/UpsertProductAssociationsIntegration.php:215]: private (and please move it to the end of the file)

**Review by mmetayer — APPROVED** [review]

**Pull request merged**